### ISSUE TITLE
How to calculate D-MPNN features only and save them

### ISSUE THREAD

**Issue by sachin-vish91** [chemprop/chemprop#167]
In the case of big files where the same molecules keep repeating multiple time. it is time-consuming to re-calculate the same features multiple time. is there any way to calculate the features and save them, then load and map them with the main dataset? for building model?

Especially when we work with cell line data the same molecule keep on repeating for different cell lines. could you help me regarding this issue?

for example for **50,000** molecules multiplied by **150** cell line, it is huge and time-consuming to re-calculate the **D-MPNN** features for the same molecules for different cell lines. It will be better if we could calculate the features once and map them with all the cell line and finally build the model.

**Comment by hesther** [comment]
Dear @sachin-vish91, you could indeed train one model on a dummy dataset, use it to predict the D-MPNN hidden representations for all your data and save them to file. Before I show you how to do that (which I think might not be the correct solution for your problem) I want to offer a few other options, and raise some questions though:

It seems like you have one or more targets for a number of molecules and cell lines. I guess that you currently add in the cell line information via an additional molecular feature, e.g. one-hot encoded? E.g.:
```
Smiles1 and Cell line 1; Target1
Smiles1 and Cell line2; Target2
Smiles2 and Cell line3; Target3 , ... 
```

You could instead rephrase your problem as a multi-task prediction:
```
Smiles 1, Target with Cell line 1, Target with Cell line2,
Smiles 2, Target with Cell line 1, Target with Cell line2, 
... (if you don't have all targets for all cell lines you can leave those fields blank)
```

This will lead to a larger model (since it now outputs several targets), but it can use the mutual information between different cell lines (which might or might not make sense depending on your data).

Or, you could simply train your model just as you do now, although it is time consuming. In fact, chemprop doesn't "re-calculate the same features multiple time" if you have the same molecule with a different additional feature, since the model changes and updates each training epoch. During training, as the model sees new cell lines (or new data in general) it adapts not only the FFN part, but also the MPNN part, which is desirable (despite time consuming) since the MPNN has trainable parameters. That is, the output of the MPNN changes and adapts with your data. Simply calculating it once and using the same representation for your model will only make sense if you find some data that you can meaningfully train the MPNN on. If you had such data, you could:

`python train.py [....all the arguments you want to use, including the (dummy) dataset you want to initially train on] --save_dir <directory>`

which would train an initial model and save it to file (MPNN and FFN). We will only reuse the MPNN part in the next step, all the rest of the model is never used again. Now you could run:

`python fingerprint.py --test_path <path to all your data> --preds_path <file to save MPNN hidden rep to> --checkpoint_dir <directory>`

which creates a csv file with your MPNN hidden representations. This will not work for ensembles of models, just a single model. You could use this file + the information from the cell lines to now train a simple feed forward neural net (with the --features_only option).

This way you would produce some fixed molecular fingerprint (with the MPNN trained on the dummy dataset) on which you then train another model. However, depending on the quality of the dummy dataset, it might me more promising to simply use one of the standard fingerprints (such as Morgan fingerprints) instead ... so I doubt this will be a good idea for most datasets.

I hope this helps!

**Comment by sachin-vish91** [comment]
Hi, @hesther 
Thank you so much for the reply. I will implement your suggestions and get back to you.

**Comment by sachin-vish91** [comment]
Hi @hesther , 

I thought about your suggestions but there one simple doubt. My data set is like
This is regression data

SMILE 1 features + cell line 1 features     target 1
SMILE 1 features + cell line 2 features     target 2
SMILE 1 features + cell line 3 features     target 3
SMILE 1 features + cell line 4 features     target 4

SMILE 2 features + cell line 1 features     target 1
SMILE 2 features + cell line 2 features     target 2
SMILE 2 features + cell line 3 features     target 3
SMILE 2 features + cell line 4 features     target 4

here i need to provide the cell line features also that i have already calculated, i just need to merge cell line features with all the molecules. Even if i use multi-task prediction as you suggested how i will supply the cell line features (160 cell line per molecule) because there is only option for the molecule feature in chemprop (to be provided in same order as smiles) not for cell line features that is why i wanted to save the D-MPNN features in .csv file. 
Please let me know if i missed something?

**Comment by sachin-vish91** [comment]
Another problem with the second suggestion. I ran the below command to generate features and built the model this worked perfectly

`chemprop_train --data_path clintox.csv --dataset_type classification --save_dir /home/vishwakarma/Documents/D-MPNN/dir`

Next when I am trying to run the next command as you suggested in your last reply

`chemprop_fingerprint --test_path clin_test.csv --checkpoint_path /home/vishwakarma/Documents/D-MPNN/dir/fold_0/model_0/model.pt --preds_path Molecules_fingerprints.csv`

see below error message:
**chemprop_fingerprint: command not found**

To be clear I did not install it from the source. that is why I used **chemprop_fingerprint** instead of **python fingerprints.py**

could you please suggest me what went wrong?

**Comment by hesther** [comment]
I see, I wrongly assumed your cell lines were only one-hot encoded. If you have featurized them, option 1 (reframing to a multitask model) will not work, sorry! 

For option 2 (just training with the whole data): How did you train your model so far? I assumed you provided the cell line features as molecule feature in Chemprop? This option is named "molecule feature" because these features are appended to the molecular hidden representation after message passing (as opposed to atomic features added before message passing), but they do not need to be the properties of a molecule. Users frequently use the molecular features to encode custom information, such as a solvent, environment, etc. So, you can provide your cell line features via the --features_path option! You would have to arrange your data like:
```
Smiles 1, target
Smiles 1, target
Smiles 1, target
...
Smiles 2, target
....
```
and your cell line features as
```
cell line 1 features
cell line 2 features
cell line 3 features
...
cell line 1 features
cell line 2 features
...
```

For saving the MPNN features: chemprop_fingerprint is a very recent addition, you will have to redownload the latest version (if your current version is more than a week or two old). The change has been in the source code for quite a while, but was only released in the precompiled version about a week ago.

**Comment by sachin-vish91** [comment]
Hi @hesther,

Thanks, Option 2 I can use. this will work perfectly. Nice explanation.

for the **chemprop_fingerprint** I am able to use now after re-installation.

But for the saving MPNN hidden representation in .csv file. as you suggested to use a dummy dataset. can I use here my main data set (50K molecules) as a dummy and learn the representation using **chemprop_train**

and then save the hidden representation using **chemprop_fingerprint** into a .csv file (for the same 50K molecules), as my goal is to get the MPNN features for my molecules?

is that be the right way or the hidden representation will be wrong? that I can use for the further model building?

I completely understand the adaptation of MPNN features with new data.

I sincerely appreciate the assistance.

**Comment by hesther** [comment]
I think saving the MPNN hidden representation would only make sense if you had a really large and good dummy dataset. For your 50K molecules, do you have any targets (without the cell line features) you could train on? Or some related data that might teach the model to focus on the same parts of the molecules? What you are trying to do (training the MPNN on one dataset, then using it as fixed representation for another model) is called "Transfer Learning", and usually only works well if the tasks you initially train on are quite related to what you want to train on in the end. Then that would be an option. 

However, if there is any way you can do the full training (option 2), I would recommend that, your MPNN features will be much better!

**Comment by sachin-vish91** [comment]
Thank you so much. yes, you are right. I will apply **option 2** because, in my dataset, there is no fixed target (it is per cell line only for multi-task learning). As MPNN will focus on the target it is better to directly use the whole dataset to train and learn the MPNN representation. **As you already mentioned that it will not re-calculate the feature for the same molecule for this case shown below right ?**

```
Smiles 1, target
Smiles 1, target
Smiles 1, target
...
Smiles 2, target
```

and cell line feature

```
cell line 1 features
cell line 2 features
cell line 3 features
...
cell line 1 features
cell line 2 features
```

**Comment by hesther** [comment]
For each training batch, Chemprop will do a forward pass through the MPNN and FFN, compute the loss of the predictions compared to the true targets, backpropagate that loss and update the parameters of both MPNN and FFN. I am not quite sure whether with "not re-calculate the MPNN features" you mean it will not compute something again and again which is the same (true - because it is not the same features, they change with each parameter update) or if you mean that Chemprop will reuse old MPNN features if it already has seen a specific smiles string (wrong, it will recalculate it at each datapoint, but the MPNN parameters will be different, so it must be recalculated). Does that answer your question? In other words, you will train on 50,000 * 160 = 8M datapoints, and there will be no difference in training speed whether those 8M points would contain 8M smiles * 1 cell line or 50K smiles * 160 cell lines or even 1 smiles * 8M cell lines because each training batch loss needs to be backpropagated through the whole model (MPNN and FFN)

**Comment by sachin-vish91** [comment]
Thank you so much for clearing my doubts, that's clear now.

**Comment by hesther** [comment]
You are welcome! I am closing this issue, but feel free to reopen, or open a new one if you have any more questions!
